[Chua Yi Fang] iP

onefangg opened the pull request:
# Karen
> Relentlessly judge others, and let others relentlessly judge you.

Karen will manage your tasks for you and make you work even harder.

## Features
* Will yell at you *(scary)*
* Will call your manager *(sob)*
* Will demand to see your manager's manager *(crying)*

## Download

Download the release [here](https://github.com/onefangg/ip/releases/tag/A-Jar).

## Usage (List of Commands)

1. list 
2. todo/deadline/event 
3. mark/unmark 
4. delete
5. find
6. bye :scream:

## Code Logic Flow
```
public static void main(String[] args) {
        Ui ui = new Ui();
        Parser parser = new Parser();
        Storage storage = new Storage(ui);
        boolean isExit = false;

        ui.showWelcome();
        while (!isExit) {
            try {
                String fullInput = ui.readInput();
                Command cmd = parser.parseInput(fullInput);
                cmd.execute(ui, storage);
                isExit = cmd.isExit();
            } catch (KarenException err) {
                ui.displayUserInput(err.message);
            }
        }
```

- [x] Ui, Parser and Storage instantiated
- [x] Ui reads in input
- [x] Parser translates input into Commands
- [ ] Command executes (may or may not interact with Storage)
- [ ] Continues until command 'bye'

Commit:
Add Gradle support

onefangg committed:
Implement level 1 - greet, echo, exit

onefangg committed:
Implemented storing of tasks into simple array

onefangg committed:
Implemented function to modify done status of item based on index

onefangg committed:
Implemented and added different types of Tasks

onefangg committed:
Implemented validation checks and rudimentary exception handling on input params from cli

onefangg committed:
Removed unnecessary debugging print

onefangg committed:
Added ability to delete Tasks and some superificial handling of error with regard to indexing

onefangg committed:
Added documnetation for Karen.java

onefangg committed:
Adding test cases

onefangg committed:
Refactor CLI Validation into Enum Types + Updating test cases

onefangg committed:
Saving session data and refactoring to abstract class for Task

onefangg committed:
Changed /by and /at to LocalDate type and validation for date commands

onefangg committed:
Revert "Changed /by and /at to LocalDate type and validation for date commands"

This reverts commit b56b6457e2d3fde1b5e3030d4135f2117cbe7bc1.

onefangg committed:
Deadline and Event accepts /by or /at in appropriate datetime format

onefangg committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Task.java

onefangg committed:
Cleanup issues resulting from merge conflict

onefangg committed:
Update test cases

onefangg committed:
Refactor code to Ui, Parser, Storage and Command class

onefangg committed:
Remove redundant CommandType
* Functionality replaced by Parser

onefangg committed:
Organise into 'command' and 'task' packages

onefangg committed:
Update test scripts to run with packages

onefangg committed:
Add test cases for Parser and UI

onefangg committed:
Create JAR file for release

onefangg committed:
Add minimal header comments to classes

onefangg committed:
Merge branch 'branch-A-JavaDoc'

onefangg committed:
Refactor code to follow coding standard

onefangg committed:
Implement functionality to find tasks within Storage
- Find is based on strict matching

onefangg committed:
Add header comments
- Updated comments with modified methods
- For newly added methods for find functionality

onefangg committed:
Add additional test cases for find command

onefangg committed:
Merge branch 'branch-A-CodingStandard'

onefangg committed:
Merge branch 'Level-9'

# Conflicts:
#	src/main/java/karen/Parser.java
#	src/main/java/karen/task/Task.java

onefangg committed:
Resolve missing code from merge conflict with Level-9
- Add missing switch case

onefangg committed:
Change name and parameters scope from formatIndexItem to formatIndexTask

"Item" is too generic and to standardise with other functions in the same class

onefangg committed:
Update saving tasks to local directory after each modification to the storage
* Add and delete are followed by a full save of current taskList to local directory
* Remove hard fix to save tasks from Karen App
* Add consistent throwing of errors

onefangg committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2122S2/ip

onefangg committed:
Add gradle support

onefangg committed:
Fix bug where marking and unmarking tasks is not saved to storage immediately

onefangg committed:
Implement porting of features from CLI interface to GUI

onefangg committed:
Add welcome message at application startup

onefangg committed:
Remove and clean unnecessary functions leftover from the implementation of a GUI
- Remove UI's variable to no longer requiring System.in and System.out
- Ui is to mainly centralise formatting to within the class itself
- Remove unnecessary imports

onefangg committed:
Merge branch 'branch-Level-10'

onefangg committed:
Add checkstyle support

onefangg committed:
Clean up according to checkstyle
- Fix lexicographical sorting of imports
- Modify access levels of variables and added necessary getter functions
- Fix whitespace and bracing formatting

onefangg committed:
Add custom stylings to GUI
- Add custom images
- Add color stylings to background and chat dialog

lchokhoe commented on src/main/java/karen/Parser.java:
```diff
@@ -0,0 +1,223 @@
+     * @return Command object
+     */
+    private Command createCommand(String fullInput) {
+        // extract first word
+        final Matcher matcher = KEYWORD_FORMAT.matcher(fullInput);
+        if (!matcher.matches()) {
+            return new InvalidCommand(NA_MESSAGE);
+        }
+
+        String keyWord = matcher.group("keyWord");
+        Command getCommand;
+
+        switch (keyWord) {
+        case "list":
+            getCommand = new ListCommand();
+            break;
+        case "bye":
+            getCommand = new ByeCommand();
+            break;
+        case "todo":
+        case "deadline":
+        case "event":
+            getCommand = prepareAdd(keyWord, fullInput);
+            break;
+        case "mark":
+        case "unmark":
+            getCommand = prepareModify(keyWord, fullInput);
+            break;
+        case "delete":
+            getCommand = prepareDelete(keyWord, fullInput);
+            break;
+        case "find":
+            getCommand = prepareFind(keyWord, fullInput);
+            break;
+        default:
+            getCommand = new InvalidCommand(NA_MESSAGE);
+            break;
+        };
+        return getCommand;
+    }
```
Great use of switch-case statement to account for the keywords Karen recognizes 👍 
Perhaps Karen could be more helpful to her users when they enter an InvalidCommand by listing the Commands that she understands? (list, bye, todo etc.) 😅

onefangg replied:
Thanks, that's a great point I overlooked

lchokhoe commented on src/main/java/karen/command/ByeCommand.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public String execute(Ui ui, Storage storage) {
+        return Ui.GOODBYE;
```
A really succinct way to handle the 'bye' command 👍 
Would it be a good idea to also consider a case where user inputs 'bye [arg]' accidentally? 

_Perhaps the user mistakenly types 'bye' when he intended to type 'todo' with an argument 😆 in which case Karen can be helpful by reminding him/her that 'bye' takes no arguments and maybe he/she should type 'bye' again._

lchokhoe commented on src/main/java/karen/task/Task.java:
```diff
@@ -0,0 +1,69 @@
+     * for writing to file
+     * @return String representation of attributes or data related to Task for storage
+     */
+    public abstract String toSaveData();
```
Great use of abstract method here to avoid code repetition 👍

lchokhoe reviewed (commented):
Overall I think the code was well written with a lot of attention to detail 👍 

Perhaps Karen could be a bit nicer to users who input an invalid command by pointing them in the right direction? Providing some Command suggestions would help a lot 😆

Denniszedead commented on src/main/java/karen/Main.java:
```diff
@@ -0,0 +1,50 @@
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+import javafx.stage.Stage;
+import karen.gui.MainWindow;
```
Should the two import statements be separated out?

Denniszedead commented on src/main/java/karen/command/InvalidCommand.java:
```diff
@@ -0,0 +1,53 @@
+     * @param message custom message to describe invalidity
+     */
+    public InvalidCommand(String message) {
+        this.message = message;
```
The usage of this keyword 'this' is fine as the parameter message shadows the object variable message. 👍

Denniszedead commented on src/main/java/karen/command/ListCommand.java:
```diff
@@ -0,0 +1,17 @@
+    public ListCommand() {
+    }
+
+    @Override
```
For Overidden methods, you can use {@inheritDoc} on your Java Doc method header.

onefangg replied:
Good catch - overlooked on this. Thanks for the suggestion!

Denniszedead commented on src/main/java/karen/gui/MainWindow.java:
```diff
@@ -0,0 +1,80 @@
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+import javafx.util.Duration;
+import karen.Karen;
```
Should the import statement be seperated?

Denniszedead commented on src/main/java/karen/gui/MainWindow.java:
```diff
@@ -0,0 +1,80 @@
+package karen.gui;
+
+import javafx.animation.PauseTransition;
+import javafx.application.Platform;
+import javafx.fxml.FXML;
+import javafx.scene.control.Button;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+import javafx.util.Duration;
+import karen.Karen;
+import karen.Ui;
```
Perhaps the import statement should be arranged in alphabetical order.

Denniszedead reviewed (commented):
Hi Yi Fang, great work 👍 in following the java statements.

However:
- For java docs, the description and the @parameter should be separated by an empty line.
- To increase readability, especially for long chunk of code, perhaps you can add in header comments to briefly describe what they do.

Recommendation:
- Try to use the keyword ``this`` less to prevent any overheads, unless the variable is overshadowed by the parameter. For example:
```java
String classVariable;

void foo(String entry) {
     classVariable = entry;
}
```
There is no need to use ``this.classVariable`` above.
However, in this example: 
```java
String classVariable;

void foo(String classVariable) {
     this.classVariable = classVariable;
}
```
There is a need to use ``this.classVariable`` as the ``classVariable`` in the parameter shadows the class variable ``classVariable``.

Feel free to approach me if you have any queries. All the best for CS2103!

onefangg committed:
Add custom stylings to GUI
- Add custom images
- Add color stylings to background and chat dialog

TeddYE commented on src/main/java/karen/Parser.java:
```diff
@@ -0,0 +1,223 @@
+                matcher = TODO_FORMAT.matcher(fullInput);
+                matcher.find();
+                return new AddCommand(new ToDo(matcher.group("description")));
+            } else if (keyWord.equals("deadline")) {
```
Maybe you can consider using switch case here so that it is easier to manage when more keywords might be added in the future.

TeddYE commented on src/test/java/UiTest.java:
```diff
@@ -0,0 +1,33 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class UiTest {
+    private final PrintStream standardOut = System.out;
```
Shouldn't constant names be all uppercase using underscore to separate words?

onefangg replied:
I didn't realised I overlooked the checkstyle checks for my test cases until you brought it up, thanks!

TeddYE commented on src/main/java/karen/task/Deadline.java:
```diff
@@ -0,0 +1,25 @@
+ * Stores the (/by) date in which the Deadline occurs by
+ */
+public class Deadline extends Task {
+    protected LocalDate by;
```
Maybe can use a more descriptive name for your variable such as dueDate?

TeddYE reviewed (commented):
Overall, I found your code easy to read.

onefangg committed:
Clean up test cases
- Refactored UiTest to remove testing based on print statements
- Sorted import statements

onefangg committed:
Refactor code
- Add space inbetween main message and parameters in the comments
- Replace usage of `this` in constructor and referenced areas to make code more Java-like
- Add @inheritDoc for overwritten methods under karen.task
- Add comments for execute method in karen.command
- Minor renaming variables

onefangg committed:
Add assertions

Assertions mainly added in areas to supplement logic-checking in switch cases or if-blocks where default configurations are already set in place; hence the code should not possibly be an empty or null

onefangg committed:
Refactor code and replace configurations
- Remove additional this statements to make code more Java-like
- Add minor comments and edits to spacing
- Replace references to static configs and launch config

onefangg committed:
Remove function in Storage to display warning with Ui

The chatbot's I/O handling has changed from CLI to GUI. This means that Storage objects' interaction with Ui will no longer be displayed on the GUI.

All references in Storage to Ui are removed.

This is done in preparation to prevent future conflicts if and when Storage needs to propagate message towards the GUI instead.

onefangg committed:
Refactor Code
- Remove empty constructors
- Remove deprecated methods
- Remove duplicate variables in different classes

onefangg committed:
Replace error logging messages used with InvalidCommand and KarenException to enum

Magic strings used to be scattered throughout the codebase. This makes it hard to track the error messages logged to user and hard to configure.

Hence, InvalidMessage acts as a config to store all error messages logged to user.

With the change, it is easier to track error logging messages in a single enum and allows for standardisation of messages.

onefangg committed:
Replace `if` block with `switch` block for consistency

onefangg committed:
Merge branch 'branch-A-CodeQuality'

onefangg committed:
Merge branch 'master' into branch-A-Assertions

onefangg committed:
Merge pull request #1 from onefangg/branch-A-Assertions

Branch a assertions

onefangg committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	src/main/java/karen/Parser.java

onefangg committed:
Merge branch 'master' of https://github.com/onefangg/ip

# Conflicts:
#	src/main/java/karen/Parser.java

onefangg committed:
Implement C-Update

Added to allow edits to Tasks stored in Karen

Add functionality to edit `description` attribute of Task stored.

This is to allow user to edit Tasks without deleting and recreating Task object

onefangg committed:
Merge pull request #2 from onefangg/branch-C-Update

Implement C-Update

onefangg committed:
Add CI Integration

onefangg committed:
Fix Github Actions directory pathing

onefangg committed:
Add fixes for successful gradle build
- Removed configs for old testing redirection scripts
- Fixed checkstyle issues
- Fixed UiTest

onefangg committed:
Add time attribute to creation of Deadline and Event

onefangg committed:
Merge branch 'branch-BugFixes'

onefangg committed:
Refactor code to update error logging/handling with InvalidMessage Enum

onefangg committed:
Add test cases and minor patches
- Add test cases for ToDo, Event, Deadline
- Update test cases for Ui
- Clean up comments, add credits for adaption, and removed unneccessary variables

onefangg committed:
Update comments to follow standard conventions

onefangg committed:
Add product screenshot

onefangg committed:
Add docs to display on User Guide for website

onefangg committed:
Fix bug with Storage
- NPE thrown due to BufferedReader not initialised when there is not ./data directory. Silently resume execution as Karen will continue on with an empty ArrayList
